ElasticSearch 5 support

AlmogBaku opened the pull request:
Solves #1410

This PR finally add the support for ElasticSearch 5 (with backward compatibility for ES2)

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/heapster/1607)
<!-- Reviewable:end -->

AlmogBaku commented:
/assign @AlmogBaku @andyxning @huangyuqi

AlmogBaku commented:
We need to write tests for ES5 and ES4

AlmogBaku commented:
@k8s-bot ok to test

AlmogBaku commented:
@k8s-bot test this

AlmogBaku commented:
cc @rikatz

AlmogBaku commented:
@k8s-ci-robot test this

AlmogBaku commented:
@DirectXMan12 I don't understand why the tests fails.. it's not even related to the PR

AlmogBaku commented:
@rikatz can you help us out with reviewing this?

rikatz commented:
@AlmogBaku sure. Will review tomorrow :)

AlmogBaku committed:
ElasticSearch: Add supports for Ignest pipelines

This commit adds support for ElasticSearch ingest pipelines.  This
only works on ES5+.

AlmogBaku commented:
I added a support for Ignest pipeline.

AlmogBaku commented:
@k8s-bot test this

AlmogBaku commented:
@k8s-bot test this

rikatz commented:
@AlmogBaku sorry for taking so long. I've faced the same issue while coding this, but when testing here locally this doesn't happens. It seems it's related with some kind of timeout. 

Have you tried to (only as a test) change lines 55 and 56 from manager_test.go from 3 to 2 to see what happens (even with the other sinks)?

rikatz commented:
Sorry @AlmogBaku I've been pretty busy in work. Have seen that this worked now. Will try to test the code tomorrow and give you a feedback :)

AlmogBaku commented:
@rikatz any update?

rikatz commented:
@AlmogBaku sorry. I'll be testing this today and will return you in the end of this day!

rikatz commented:
@AlmogBaku while running heapster with elasticsearch 5, it gaves me the following error:

``
E0510 15:30:33.101740   12777 factory.go:84] Failed to create sink: Failed to create ElasticSearch client: Failed to an ElasticSearch Client: no Elasticsearch node available``

I'm using the following directive: ``--sink="elasticsearch:?nodes=http://ipaddress:9200&index=testEvent&ver=5" ``

The Elasticsearch is available and answering:

```
curl http://ipaddress:9200
{
  "name" : "Pz3qaoM",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "BUd5OBEGTpCIE6Xak5qglA",
  "version" : {
    "number" : "5.4.0",
    "build_hash" : "780f8c4",
    "build_date" : "2017-04-28T17:43:27.229Z",
    "build_snapshot" : false,
    "lucene_version" : "6.5.0"
  },
  "tagline" : "You Know, for Search"
}
```
Am I missing something?

rikatz commented:
@AlmogBaku some more info:

I'm running a standalone Elasticsearch, and started it now with 'sniff=false' (as in docs). Running with ver, disabling health check, all have the same result.

The following is a dump from communication between Heapster and Elastic:

```
HEAD / HTTP/1.1
Host: 10.X.X.X:9200
User-Agent: Go-http-client/1.1

HTTP/1.1 200 OK
content-type: application/json; charset=UTF-8
content-length: 327

GET /_nodes/http HTTP/1.1
Host: 10.X.X.X:9200
User-Agent: elastic/5.0.36 (linux-amd64)
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip

HTTP/1.1 200 OK
content-type: application/json; charset=UTF-8
content-encoding: gzip
transfer-encoding: chunked
```

AlmogBaku commented:
@rikatz 
1) please double check your config
2) please send me logs with verbose enabled (as described here https://github.com/kubernetes/heapster/blob/master/docs/debugging.md )

AlmogBaku commented:
@rikatz ?

chancez commented:
@AlmogBaku I for what it's worth, I got this working just fine. I had to make a small addition to the PR to add an option to disable SSL verification on the http.Client since I'm using a self signed TLS cert for ES (using the elasticsearch-operator) and didn't bother with getting the certs correctly mounted in the eventer container, but otherwise it works great.

Here's the manifest I'm using: https://gist.github.com/e725dea50300bafc781382546bfe6f2b

The additional code is just this:
```
if len(opts["noVerifySSL"]) > 0 {
	httpClient := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{
				InsecureSkipVerify: true,
			},
		},
	}
	startupFnsV2 = append(startupFnsV2, elastic2.SetHttpClient(httpClient))
	startupFnsV5 = append(startupFnsV5, elastic5.SetHttpClient(httpClient))
}
```

AlmogBaku commented:
@chancez I guess that's belong to a different PR..

chancez commented:
@AlmogBaku Yeah, I'm waiting for this one to get merged and I'll open a PR for it. Mostly wanted to say this is working great perfectly for me.

rikatz commented:
Sorry!!! I've been pretty busy at work :( will test again on monday and will let you know!

Ricardo P. Katz

> Em 20 de mai de 2017, às 14:22, Almog Baku <notifications@github.com> escreveu:
> 
> @rikatz ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rikatz commented:
@chancez Can you please quickly write how you started ES5 and Heapster?

Thanks!

chancez commented:
@rikatz I'm using https://github.com/upmc-enterprises/elasticsearch-operator and the manifest I'm using is in the gist above. Just know I've made a minor modification to this code to work with my elasticsearch, as mentioned, so I can disable SSL verification.

rikatz commented:
OK. I'm testing this again now and will let you know

rikatz commented:
@AlmogBaku tested here in ES5 and it worked really fine :)

LGTM!

Thanks and sorry about delaying this so much :(

rikatz commented:
I was just wondering here if it wouldn't be better to have a 'Generic Timestamp' (like a @timestamp) for each metric, as this breaks each Timestamp into CpuMetricsTimestamp, MemoryMetricsTimestamp and this, in kibana makes a lot difficult to create dashboards (as each metric have it's own timestamp and not a generic one).

Also, after pushing some metrics I've faced some 'Failed to push data to sink: ElasticSearch Sink' but don't know why. Heapster is pushing metrics anyway, so probably is something specific to some object inside my cluster.

rikatz commented:
@AlmogBaku other 'review'. While creating a dashboard in Grafana, I was trying to aggregate metrics per namespace and per pod-name.

When using MetricsTags.pod_id it worked fine, but when using MetricsTags.pod_name I've faced the error related to [fielddata](https://www.elastic.co/guide/en/elasticsearch/reference/current/fielddata.html)

Don't know if this is fine for now, but probably we need a better field mapping in this Sink, as those searches/aggregations could be used :)

Anything I can help with?

SupreethKadalur commented:
Hello @AlmogBaku , @rikatz - Please let us know if ES5 PR is merged and ready for use. I have the same problem and unable to send data from heapster v 1.3.0 to ES5

@rikatz - I totally agree that we should have  @timestamp for each metrics :-) this will solve many problems I have while creating dashboards in kibana.

DirectXMan12 commented:
Let me know when you guys are ready to merge this one, and I will.

rikatz commented:
@AlmogBaku I'm fine, for now if we just create a generic timestamp for all metrics :)

In the (near) future we can make a template mapping of the metrics that is created before the first insertion in the new index, but this is something that can be dealled in another PR.

bsafwen commented:
Hi, in my case, using your  commit  a68f4ac, heapster is able to find elasticsearch, I am able to retrieve most of the data but my problem is that  although I see a cpu/usage_rate in the mappings, I cannot see it in my queries which means the field has no data. On the other hand, cpu/usage is correct.

AlmogBaku commented:
@rikatz 
1) that's what we do right now- we create the mapping in advance.
2) I think the reason we created different timestamps was to allow you to create different "types"/"index patterns" in the kibana dashboard.. because otherwise you just can't see the other fields
take a look at #1313 
Please open a new ticket, since it's not related to this

@bsafwen it's also related to my comment above, you should configure the different types..

@chancez personally, I don't think you should disable it in any case.. you should just add your self-signed CA.. anyway - please open a new issue



It seems to me like it's okay to merge now @DirectXMan12

bsafwen commented:
@AlmogBaku  Thank you for your reply. I changed the index from heapster-2016-06-09 to heapster* and it worked, still trying to figure why.

AlmogBaku commented:
because the index is with date pattern....

SupreethKadalur commented:
@AlmogBaku, @DirectXMan12  - I'm using image "**docker pull kubernetes/heapster**" from docker hub. Merging and building this PR (heapster to Elastic 5) will push the image to docker hub ? Please let me know otherwise. Thank you

AlmogBaku commented:
/assign @DirectXMan12 @piosz

DirectXMan12 commented:
@AlmogBaku could you squash some of the "gofmt" and "fix" commits?  Each commit should be one logical unit of features (or an update to Godeps).  After that, I'll merge.  Thanks for the hard work!

DirectXMan12 commented:
I'll squash myself so we can get this in the upcoming release

DirectXMan12 commented:
Alright, squashed into three commits (Godeps, ES5, ingest).  Should be good to merge once it passes the tests.

AlmogBaku committed:
ElasticSearch: Add ES5 Support

This commit adds in support for ES5 to the ElasticSearch sink.
The ES5 support lands side-by-side with the ES2 support, since
the APIs are slightly different.

AlmogBaku committed:
Update Godeps for ES5 support

This commit updates the Godeps for ES5 support in the ElasticSearch
sink.

DirectXMan12 commented:
/retest

DirectXMan12 commented:
looks like a flake unrelated to this PR

/retest

AlmogBaku commented:
/retest

DirectXMan12 commented:
looks good, merging

rikatz commented:
@AlmogBaku thanks!! :)

DirectXMan12 commented:
Thanks everyone for all their hard work on this :-)

AlmogBaku commented:
Hallelujah!

On Fri, Jun 23, 2017 at 9:51 PM Solly Ross <notifications@github.com> wrote:

> Thanks everyone for all their hard work on this :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/heapster/pull/1607#issuecomment-310745318>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGCpoaNcw6wBGcbA2uZMuRjeWtjAbafks5sHAkegaJpZM4NFmG0>
> .
>
-- 

<http://www.rimoto.net/>

www.rimoto.com <http://www.rimoto.net/>


*Almog Baku*


*CTO & Cofounder *
Mobile: +972.50.2288.744
Social:  * <http://www.facebook.com/AlmogBaku>
<http://www.linkedin.com/in/almogbaku>*

SupreethKadalur commented:
@DirectXMan12 , @AlmogBaku - I see that ES5 fix is merged and Heapster v1.4.0-beta.0 is created. I'm now using docker image "gcr.io/google_containers/heapster-amd64:v1.4.0-beta.0" but still see the same problem I had with heapster-amd64:v1.3.0 image. Am I missing something here? Please suggest. Thanks

chancez commented:
@SupreethKadalur This got merged after that, you'll have to wait for another pre-release.

SupreethKadalur commented:
Thank you all for the support and fix. Now ES5 issue is fixed with heapster v1.4.0

piosz commented:
Yes, this is included in 1.4.0 as described in  https://github.com/kubernetes/heapster/releases/tag/v1.4.0
